Added more configuration options to ProductionAirdrop and made the CHOAM space frigate come in from the closest map edge cell

penev92 opened the pull request:
Inspired by #17825 and now supersedes it.
Unfortunately #16595 wasn't doing the job it needed to do for this (and I still find its behaviour questionable).
This could also open a bunch of interesting options for the TD Airstrip (although I doubt they'll be accepted) and for third-party mods when one plays around with the new controls.

The last commit makes the CHOAM frigate in D2k come from the closest map edge and continue in a straight line after unloading its cargo.

(_P.S.: Admittedly today wasn't a great name for naming things._)

penev92 committed:
Moved some info fetching to the constructor in ProductionAirdrop

penev92 committed:
Added DeliveryActorPathInfo and some minor refactoring to ProductionAirdrop

penev92 committed:
Added an option to ProductionAirdrop to spawn the delivery aircraft at the closest map edge cell

penev92 committed:
Added SpawnOffset for the delivery actor in ProductionAirdrop

penev92 committed:
Refactored ProductionAirdrop delivery actor facing

penev92 committed:
Renamed ProductionAirdropInfo.Facing to LandingFacing to better reflect its purpose

penev92 committed:
Added configurable delivery actor exits to ProductionAirdrop

# Conflicts:
#	OpenRA.Mods.Common/Traits/Buildings/ProductionAirdrop.cs

penev92 committed:
Made ProductionAirdrop public

penev92 committed:
Added and ran an upgrade rule for the ProductionAirdrop rework

penev92 committed:
Set up ProductionAirdrop on the D2k Starport to better match the original game's behaviour

abcdefg30 commented on OpenRA.Mods.Common/UpdateRules/Rules/20200202/ImproveProductionAirdrop.cs:
```diff
@@ -0,0 +1,47 @@
+﻿using System.Collections.Generic;
```
Missing header.

penev92 replied:
Added.

Mailaender commented:
`ExitType: SameAsEntry` has some weird behavior: https://gfycat.com/miserlyslowfruitfly

penev92 commented:
The bug shown in the gif is a result of crappy interaction between sliding aircraft, low turn speed, the inexplicable presence of `FlyTimed(-1)` as a child activity of `FlyOffMap` and the Starport being close enough to the map's edge to trigger.
If the Starport is further from the edge the frigate has time to normalize its facing.
If the frigate's turn speed is increased it has time to normalize its facing.
But if it can't normalize it, then it flies in a straight line with an unwanted facing (because it `CanSlide`) and it reaches its destination with a random facing. Then flies in a straight line in that direction because of `FlyTimed(-1)`.

penev92 commented:
Rebased and added the missing header.
I can fix the reported bug the easy way - by increasing the frigate's turn rate (although the bug isn't even related to the PR), but I would like to discuss a better (proper) fix, since it is a result of several layers of issues and taking the easy way out sounds cheap (and will leave the issues to eventually be encountered by someone else).

Mailaender reviewed (changes requested):
~~Can you remove the conflicts comment?~~ You already did. I forgot to refresh.

Mailaender commented on OpenRA.Mods.Common/Traits/Buildings/ProductionAirdrop.cs:
```diff
@@ -112,11 +103,142 @@ public override bool Produce(Actor self, ActorInfo producee, string productionTy
+			var owner = self.Owner;
+			var map = owner.World.Map;
+			var bounds = map.Bounds;
+			var mpStart = owner.World.WorldActor.TraitOrDefault<MPStartLocations>();
```
Can this be cached as well?

Mailaender commented:
Works as promised. :+1:

Inq8 commented:
Works as intended.  One small criticism would be the loop a Non-VTOL delivery aircraft does if the ProductionAirdrop actor is near the edge.
![edgedelay](https://user-images.githubusercontent.com/42888724/81267602-a80cba80-903e-11ea-8e4c-39da23aebbc4.gif)

It would be nice if it could just come straight into land after production or use a distance-scaling production delay so it is held for a certain time period & then comes straight into land.

With that said though I still think the TD mod would benefit from using  DropSiteClosestEdge Entry & StraightAhead Exit even as is.

tovl commented:
>  One small criticism would be the loop a Non-VTOL delivery aircraft does if the ProductionAirdrop actor is near the edge.

The problem here is that the aircraft needs a minimal distance from the landing spot to descend from cruise altitude, but the aircraft spawns right in front of the airfield so it has to loop around first to get at the right distance. This can be solved by letting the aircraft spawn a fixed distance away from the airfield, just like we do for the fixed approach.

tovl commented on OpenRA.Mods.Common/Traits/Buildings/ProductionAirdrop.cs:
```diff
@@ -12,13 +12,16 @@
 
 namespace OpenRA.Mods.Common.Traits
 {
+	public enum EntryType { Fixed, PlayerSpawnClosestEdge, DropSiteClosestEdge }
```
As clarified in commit 75a0cb44222d7be5e610bebd62cf99cdd3574cb0 `PlayerSpawnClosestEdge` isn't really spawning at the closest edge per se, neither was that the original intent. This name should probably be fixed. It would also be good if there was a proper description for all the different options here and for `ExitType`.

penev92 replied:
Ah that makes sense. Will change.

penev92 replied:
Scratch that. I don't understand what it's doing. Can you give me a name and a description for it and I'll add them along with descriptions for the rest?

tovl replied:
Alright, how about: `PlayerBaseline` "Spawns at the map edge directly behind the player spawn point (from the perspective of the map center)"

A bit of context: The idea here was inspired by some tabletop games where special significance is given to the edge of the board that is physically closest to the player. Sometimes called a baseline or home-row  or something like that. In those games reinforcements can only be placed at your map edge (or sometimes units must exit through the _opponent's_ baseline to gain an advantage, like checkers).

In order to make the concept applicable to the wide variety of possible spawn point configurations (including FFAs) the baselines are defined radially, like pizza slices. This means that if opponents spawn in opposing corners, the reinforcements will arrive along the diagonal of the map and in Tiberium Garden type maps, the reinforcement locations are appropriately separated.

penev92 replied:
I think I get the idea, but testing on the map Arcade threw me off. Thanks. I'll try to update along with everything else shortly.

tovl commented on OpenRA.Mods.Common/Traits/Buildings/ProductionAirdrop.cs:
```diff
@@ -112,11 +103,142 @@ public override bool Produce(Actor self, ActorInfo producee, string productionTy
+				actor.QueueActivity(new CallFunc(() =>
+				{
+					var exitPosition = deliveryPath.CalculateExitPosition(actor);
+					actor.QueueActivity(new FlyOffMap(actor, Target.FromCell(w, exitPosition)));
```
Queueing a root-level activity inside of a `CallFunc` seems a bit fishy to me. If anything, these should be childactivities, but it would be better if we calculate `exitPosition` ahead of time and avoid the `CallFunc` altogether.

tovl replied:
Since `exitPosition` is always defined, there seems to be no need for `FlyOffMap` here. We can just queue a regular `Fly`.

tovl commented on OpenRA.Mods.Common/Traits/Buildings/ProductionAirdrop.cs:
```diff
@@ -112,11 +103,142 @@ public override bool Produce(Actor self, ActorInfo producee, string productionTy
+				case ExitType.Closest:
+					exitPositionFunc = actor =>
+					{
+						// By the time the func gets executed the map bounds could have changed, so fetching them fresh.
```
Can they?

penev92 replied:
Currently no, but that is a prerequisite for TS missions, so I assume we'll get it eventually in time for the TS Public Alpha.

abcdefg30 replied:
That feature should really come with an INotify callback, so I don't think we should calculate this here, or make assumptions about future code at this moment.

penev92 replied:
Fair enough on the `INotify`.

tovl commented on OpenRA.Mods.Common/Traits/Buildings/ProductionAirdrop.cs:
```diff
@@ -112,11 +103,142 @@ public override bool Produce(Actor self, ActorInfo producee, string productionTy
+
+					bounds = self.World.Map.Bounds;
+					if (info.SpawnOffset.X != 0 && startPos.X != bounds.X && startPos.X != bounds.X + bounds.Width)
+						startPos += new CVec(info.SpawnOffset.X, 0);
```
What is the purpose of `SpawnOffset` and `ExitOffset`? Shouldn't they depend on the approach direction as well? And why do they only apply to the 'closest edge' case?

abcdefg30 commented:
No activity since a while, so closing until you can continue this. (Maybe @tovl or someone else wants to take over?)
